rfc/20241021-dynamic-provider-instances.md

```diff
@@ -0,0 +1,766 @@
+
+A module author can declare that their module expects to be passed provider instances from its caller using the optional `configuration_aliases` argument in a `required_providers` entry. For example:
+
+```hcl
+terraform {
+  required_providers {
+    aws = {
+      source = "hashicorp/aws"
+      configuration_aliases = [
+        aws.foo,
+        aws.bar,
+      ]
+    }
+  }
+}
+```
+
+The above declares that this module has local provider instance addresses `aws.foo` and `aws.bar`, but which provider configuration they actually refer to is decided by the parent module. The parent module's `module` block calling this module must therefore include a `providers` argument that describes how to map from the parent module's address namespace into the child module's address namespace:
+
+```hcl
+provider "aws" {
+  alias = "use1"
+
+  region = "us-east-1"
+}
+
+provider "aws" {
+  alias = "euw2"
+
+  region = "eu-west-2"
+}
+
+module "example" {
+  source = "./example" # This is the directory containing the previous example block
+  providers = {
+    aws.foo = aws.use1
+    aws.bar = aws.euw2
+  }
+
+  # ...
+}
+```
+
+The above configuration means that `aws.foo` in the child module refers to the same provider instance as `aws.use1` refers to in the parent module.
+
+The mapping assigned to the `providers` argument is a static mapping, so it must always use the object constructor syntax with braces `{ }`; it's not valid to assign a dynamic expression that would generate a map.
+
+To build the dependency graph OpenTofu needs to bind each provider configuration address in the child module to exactly one provider configuration address in the parent module before evaluating any expressions. It's valid to use dynamic instance expressions on the right side of the mappings in the `providers` block, but as with resources all instances of a module call must refer to instances of the same provider configuration:
+
+```hcl
+provider "aws" {
+  alias    = "by_region"
+  for_each = var.aws_regions
+
+  region = each.key
+}
+
+module "example2" {
+  source = "./another-example" # This is _not_ the directory containing the earlier example
+  for_each  = var.aws_regions
+  providers = {
+    aws = aws.by_region[each.key]
+  }
+}
+```
+
+The above means that in each instance of `module.example2` the default instance of "aws" is dynamically decided by evaluating `each.key`. The resources using `provider = aws` (implicitly or explicitly) in that module all depend on the `aws.by_region` configuration from the parent module in the dependency graph, and then choose one of its instances dynamically at runtime after evaluating `each.key`.
+
+The final interesting case is passing a whole multi-instance provider configuration between modules. For this we extend the `configuration_aliases` syntax to allow declaring that the module expects to be passed a multi-instance configuration rather than a singleton configuration:
```

Comment:
It seems like a brand-new feature in the context of the initial (current) #300 implementation. Sounds pretty powerful for multi-region modules (and IMO we should implement it at some point), however, it is not a blocker for the alpha release.

Reply:
My overall intent with this document was to anticipate likely future requests, so indeed this particular thing is new but so is much of what I proposed here!

I don't think I'd consider _anything_ I wrote in this to be "a blocker for the alpha release". The only immediate question is to what extent we want to change the initial implementation to avoid blocking other features I described here. Though I would agree that this particular thing doesn't really seem to need any special allowance in the initial implementation, since the first round would consider this `aws.foo[*]` syntax _totally invalid_ anyway, leaving us free to give it a meaning in a later release without breaking compatibility.

Reply:
UPD: I missed this from the original static-eval RFC, we just skipped it for the first version as well.